File: chunlin/133_BFS.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null) {         // Null exception, forgot to check
            return null;
        }
        HashMap<UndirectedGraphNode, UndirectedGraphNode> map = new HashMap<>();
        HashSet<UndirectedGraphNode> nodeSet = new HashSet<>();
        Queue<UndirectedGraphNode> queue = new LinkedList<>();
        queue.offer(node);
        nodeSet.add(node);
        while (!queue.isEmpty()) {
            UndirectedGraphNode cur = queue.poll();
            for (UndirectedGraphNode neighbor : cur.neighbors) {
                if (!nodeSet.contains(neighbor)) {
                    queue.offer(neighbor);
                    nodeSet.add(neighbor);
                }
            }
        }
        
        for (UndirectedGraphNode cur: nodeSet) {
            map.put(cur, new UndirectedGraphNode(cur.label));
        }
        for (UndirectedGraphNode cur: nodeSet) {
            for (UndirectedGraphNode nei: cur.neighbors) {
                map.get(cur).neighbors.add(map.get(nei));
            }
        }
        
        return map.get(node);
        
    }
}

File: jenfengchang/133BFS.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
    unordered_map<UndirectedGraphNode *, bool> visited;
    unordered_map<int, UndirectedGraphNode *> map;
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if(node == NULL) return NULL;
        UndirectedGraphNode *begin = new UndirectedGraphNode(node->label);
        map[node->label] = begin;
        vector<UndirectedGraphNode*> cur;
        cur.push_back(node);
        while(!cur.empty()) {
            vector<UndirectedGraphNode*> next;
            for(auto &tmp : cur) {
                visited[tmp] = true;
                for(auto &N : tmp->neighbors) {
                    if(map[N->label] == NULL) {
                        UndirectedGraphNode *tmpnode = new UndirectedGraphNode(N->label);
                        map[N->label] = tmpnode;
                    }
                    map[tmp->label]->neighbors.push_back(map[N->label]);
                    if(!visited[N]) {
                        visited[N] = true;
                        next.push_back(N);
                    }
                }
            }
            cur = next;
        }
        return begin;
    }
};
